dcos/cosmos.py

```diff
@@ -0,0 +1,401 @@
+            if len(headers_preference) > 1:
+                # reattempt with one less item in headers_preference
+                response = self._call_cosmos(url,
+                                             http_request_type,
+                                             headers_preference[1:],
+                                             data,
+                                             json,
+                                             **kwargs)
+            else:
+                raise e
+        return response
+
+    def _get_endpoint_url(self, endpoint):
+        """
+        Gets the url for the cosmos endpoint 'endpoint'
+
+        :param endpoint: a cosmos endpoint, of the form 'x/y',
+        for example 'package/repo/add' or 'service/start'
+        :type endpoint: str
+        :return: the url of endpoint
+        :rtype: str
+        """
+        return urllib.parse.urljoin(self.cosmos_url, endpoint)
+
+    def _get_request_version_preferences(self, endpoint):
+        """
+        Gets the list of versions for endpoint in preference order.
+        The first item is most preferred, and last is least preferred.
+
+        :param endpoint: a cosmos endpoint, of the form 'x/y',
+        for example 'package/repo/add' or 'service/start'
+        :type endpoint: str
+        :return: list of versions in preference order
+        :rtype: list[str]
+        """
+        return self._request_versions.get(endpoint)
+
+    def _get_http_method(self, endpoint):
+        """
+        Gets the http method cosmos expects for the endpoint
+        :param endpoint: a cosmos endpoint, of the form 'x/y',
+        for example 'package/repo/add' or 'service/start'
+        :type endpoint: str
+        :return: http method type
+        :rtype: str
+        """
+        method = self._http_method.get(endpoint)
+
+        if method is not 'post' and method is not 'get':
+            raise DCOSException('Bad method type')
+
+        return method
+
+    def _get_header(self, endpoint, version, headers=None):
+        """
+        Given an cosmos endpoint, a version, and any extra header values,
+        gets the header that can be used to query cosmos at endpoint.
+        Any key in headers will appear in the final header. In effect the
+        user function can overwrite the default header.
+
+        :param endpoint: a cosmos endpoint, of the form 'x/y',
+        for example 'package/repo/add' or 'service/start'
+        :type endpoint: str
+        :param version: The version of the request
+        :type version: str
+        :param headers: extra keys for the header
+        :type headers: dict[str, str]
+        :return: a header that can be used to query cosmos at endpoint
+        :rtype: dict[str, str]
+        """
+        simple_header = {
+            'Content-Type': self._get_content_type(endpoint),
+            'Accept': self._get_accept(endpoint, version)
+        }
+        return remove_nones(merge_dict(simple_header, headers))
```

Comment:
Users can overwrite the default headers if they want to. For example, if I want a header that does not have a 'Content-Type'. I can just pass in a {'Content-Type': None}. Since users can do this, I would like to remove all nones.